platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/TaskSchedulerBuilder.java

```diff
@@ -0,0 +1,335 @@
+     * used.
+     *
+     * @param pool the thread pool
+     * @return this
+     */
+    @NonNull
+    public TaskSchedulerBuilder<O> withPool(@NonNull final ForkJoinPool pool) {
+        this.pool = Objects.requireNonNull(pool);
+        return this;
+    }
+
+    /**
+     * Provide a custom uncaught exception handler for this task scheduler. If none is provided then the default
+     * uncaught exception handler will be used. The default handler will write a message to the log.
+     *
+     * @param uncaughtExceptionHandler the uncaught exception handler
+     * @return this
+     */
+    @NonNull
+    public TaskSchedulerBuilder<O> withUncaughtExceptionHandler(
+            @NonNull final UncaughtExceptionHandler uncaughtExceptionHandler) {
+        this.uncaughtExceptionHandler = Objects.requireNonNull(uncaughtExceptionHandler);
+        return this;
+    }
+
+    /**
+     * Build an uncaught exception handler if one was not provided.
+     *
+     * @return the uncaught exception handler
+     */
+    @NonNull
+    private UncaughtExceptionHandler buildUncaughtExceptionHandler() {
+        if (uncaughtExceptionHandler != null) {
+            return uncaughtExceptionHandler;
+        } else {
+            return (thread, throwable) ->
+                    logger.error(EXCEPTION.getMarker(), "Uncaught exception in wire {}", name, throwable);
+        }
+    }
+
+    private record Counters(@NonNull ObjectCounter onRamp, @NonNull ObjectCounter offRamp) {}
+
+    /**
+     * Combine two counters into one.
+     *
+     * @param innerCounter the counter needed for internal implementation details, or null if not needed
+     * @param outerCounter the counter provided by the outer scope, or null if not provided
+     * @return the combined counter, or a no op counter if both are null
+     */
+    @NonNull
+    private static ObjectCounter combineCounters(
+            @Nullable final ObjectCounter innerCounter, @Nullable final ObjectCounter outerCounter) {
+        if (innerCounter == null) {
+            if (outerCounter == null) {
+                return NoOpObjectCounter.getInstance();
+            } else {
+                return outerCounter;
+            }
+        } else {
+            if (outerCounter == null) {
+                return innerCounter;
+            } else {
+                return new MultiObjectCounter(innerCounter, outerCounter);
+            }
+        }
+    }
+
+    /**
+     * Figure out which counters to use for this task scheduler (if any), constructing them if they need to be
+     * constructed.
+     */
+    @NonNull
+    private Counters buildCounters() {
+        final ObjectCounter innerCounter;
+
+        if (unhandledTaskCapacity != UNLIMITED_CAPACITY) {
+            innerCounter = new BackpressureObjectCounter(unhandledTaskCapacity, sleepDuration);
+        } else if (metricsBuilder != null && metricsBuilder.isUnhandledTaskMetricEnabled()
```

Comment:
```
        // If we need to enforce a maximum capacity, we have no choice but to use a backpressure object counter.
        // 
        // If we don't need to enforce a maximum capacity, we need to use a standard object counter if any 
        // of the following conditions are true:
        //  - we have unhandled task metrics enabled
        //  - the scheduler is concurrent and flushing is enabled. This is because the concurrent scheduler's
        //    flush implementation requires a counter that is not a no-op counter.
        //
        // In all other cases, better to use a no-op counter. Counters have overhead, and if we don't need one
        // then we shouldn't use one.
```